### PR TITLE
[MLIR] Register custom gradients with Enzyme

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Description of the Change:**  MLIR changes needed for custom gradients. Not yet integrated into the frontend.
Adds:
* ForwardOp
* ReverseOp
* ReturnOp
* CustomGradOp

The ForwardOp and ReverseOp are intended to be set from the frontend to have the calling convention from Enzyme. E.g.,

```llvmir
# The return is the tape type
gradient.forward @foo(%in : tensor<f64>, %diff : tensor<f64>, %out : tensor<f64>, %cotangent: tensor<f64>) -> (tensor<f64>, ... tensor<..>) 

gradient.reverse @bar(%in : tensor<f64>, %diff : tensor<f64>, %out : tensor<f64>, %cotangent: tensor<f64>, %tape0: tensor<f64>, ... %tapeN: tensor<f64>) 
```

The returnOp is necessary because func.return is only valid inside func.func. If there is no tape, the tape is a null pointer. Having an empty struct resulted in an error, but it might be possible to fix. There is no real disadvantage of using null pointer since it is never dereferenced.

[sc-60512]
[sc-60498]

### CODE REVIEW

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp [hunk 1]
```diff
@@ -247,6 +247,92 @@ struct BufferizeForwardOp : public OpConversionPattern<ForwardOp> {
     }
 };
 
+struct BufferizeReverseOp : public OpConversionPattern<ReverseOp> {
+    using OpConversionPattern::OpConversionPattern;
+
+    LogicalResult match(ReverseOp op) const override
+    {
+        // Only match here if we have all memref arguments and return values.
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:255]: Should check that this is empty

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp [hunk 2]
```diff
@@ -247,6 +247,92 @@ struct BufferizeForwardOp : public OpConversionPattern<ForwardOp> {
...
+    void rewrite(ReverseOp op, OpAdaptor adaptor,
+                 ConversionPatternRewriter &rewriter) const override
+    {
+        auto argc = op.getArgc();
+        auto resc = op.getResc();
+        SmallVector<Value> inputs;
+        SmallVector<Value> differentials;
+        SmallVector<Value> outputs;
+        SmallVector<Value> cotangents;
+        SmallVector<Value> residuals;
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:277]: Naming convention please.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp [hunk 3]
```diff
@@ -882,6 +882,132 @@ struct ForwardOpPattern : public ConvertOpToLLVMPattern<ForwardOp> {
...
+        }
+
+        auto ctx = rewriter.getContext();
+        // This gives me a struct of the following type
+        // { memref0, memref1, memref2, ... memrefN }
+        auto tapeTy = LLVM::LLVMStructType::getLiteral(ctx, residualStructs);
+        // This gives me { { memref, ..., memrefn } }
+        auto wrappedFlatTapeStructTy = LLVM::LLVMStructType::getLiteral(ctx, {tapeTy});
+        auto ptrTy = LLVM::LLVMPointerType::get(ctx);
+        Type retty = tapeCount > 0 ? dyn_cast<Type>(wrappedFlatTapeStructTy) : dyn_cast<Type>(ptrTy);
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: Shouldn't call it retty

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp [hunk 4]
```diff
@@ -882,6 +882,132 @@ struct ForwardOpPattern : public ConvertOpToLLVMPattern<ForwardOp> {
...
+        IRMapping map;
+
+        if (tapeCount > 0) {
+        auto loc = op.getLoc();
+        auto lastIdx = newFuncInputTys.size() - 1;
+        Value wrappedStructValAgg = func.getArgument(lastIdx);
+
+        SmallVector<Value> tapestructs;
+        for (auto i = 0; i < tapeCount; i++) {
+            SmallVector<int64_t> pos = {0, static_cast<int64_t>(i)};
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: Maybe I don't need the zero here.

File: mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp [hunk 5]
```diff
@@ -57,6 +58,17 @@ struct GradientConversionPass : impl::GradientConversionPassBase<GradientConvers
         target.addLegalDialect<arith::ArithDialect, linalg::LinalgDialect, func::FuncDialect,
                                index::IndexDialect, memref::MemRefDialect>();
 
+        // This is a bit unfortunate.
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp:61]: Maybe not that unfortunate, please verify

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp (hunk 4 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: Mmm... tried this patch:

```
diff --git a/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp b/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
index eb4802aa..c31bda2b 100644
--- a/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
+++ b/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
@@ -941,10 +941,8 @@ struct ReverseOpPattern : public ConvertOpToLLVMPattern<ReverseOp> {
         // This gives me a struct of the following type
         // { memref0, memref1, memref2, ... memrefN }
         auto tapeTy = LLVM::LLVMStructType::getLiteral(ctx, tapeStructs);
-        // This gives me { { memref, ..., memrefn } }
-        auto wrappedFlatTapeStructTy = LLVM::LLVMStructType::getLiteral(ctx, {tapeTy});
         auto ptrTy = LLVM::LLVMPointerType::get(ctx);
-        Type inputTapeTy = tapeCount > 0 ? dyn_cast<Type>(wrappedFlatTapeStructTy) : dyn_cast<Type>(ptrTy);
+        Type inputTapeTy = tapeCount > 0 ? dyn_cast<Type>(tapeTy) : dyn_cast<Type>(ptrTy);
         newFuncInputTys.push_back(inputTapeTy);
 
         auto newFuncTy = FunctionType::get(ctx, newFuncInputTys, TypeRange{});
@@ -968,7 +966,7 @@ struct ReverseOpPattern : public ConvertOpToLLVMPattern<ReverseOp> {
 
         SmallVector<Value> tapestructs;
         for (auto i = 0; i < tapeCount; i++) {
-            SmallVector<int64_t> pos = {0, static_cast<int64_t>(i)};
+            SmallVector<int64_t> pos = {static_cast<int64_t>(i)};
             Value tapeStructIth =
                 rewriter.create<LLVM::ExtractValueOp>(loc, wrappedStructValAgg, pos);
             tapestructs.push_back(tapeStructIth);
diff --git a/mlir/test/Gradient/ConversionTest.mlir b/mlir/test/Gradient/ConversionTest.mlir
index 222a61b5..e1aa2c26 100644
--- a/mlir/test/Gradient/ConversionTest.mlir
+++ b/mlir/test/Gradient/ConversionTest.mlir
@@ -92,7 +92,7 @@ module @test1 {
   }
 
   // CHECK-LABEL: func.func private @rev.rev
-  // CHECK: [[in0ptr:%.+]]: !llvm.ptr, [[diff0ptr:%.+]]: !llvm.ptr, [[out0ptr:%.+]]: !llvm.ptr, [[cotan0ptr:%.+]]: !llvm.ptr, [[tape:%.+]]: !llvm.struct<(struct<(struct<(ptr, ptr, i64)>)>)>) attributes {passthrough = ["noinline"]}
+  // CHECK: [[in0ptr:%.+]]: !llvm.ptr, [[diff0ptr:%.+]]: !llvm.ptr, [[out0ptr:%.+]]: !llvm.ptr, [[cotan0ptr:%.+]]: !llvm.ptr, [[tape:%.+]]: !llvm.struct<(struct<(ptr, ptr, i64)>)>) attributes {passthrough = ["noinline"]}
   gradient.reverse @rev.rev(%arg0: memref<f64>, %arg1: memref<f64>, %arg2: memref<f64>, %arg3: memref<f64>, %arg4: memref<f64>) attributes {argc = 1 : i64, implementation = @rev, resc = 1 : i64, tape = 1 : i64} {
 
     // CHECK: [[in0struct:%.+]] = llvm.load [[in0ptr]] : !llvm.ptr -> !llvm.struct<(ptr, ptr, i64)>
@@ -103,7 +103,7 @@ module @test1 {
     // CHECK: [[out0memref:%.+]] = builtin.unrealized_conversion_cast [[out0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
     // CHECK: [[cotan0struct:%.+]] = llvm.load [[cotan0ptr]] : !llvm.ptr -> !llvm.struct<(ptr, ptr, i64)>
     // CHECK: [[cotan0memref:%.+]] = builtin.unrealized_conversion_cast [[cotan0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
-    // CHECK: [[tape0struct:%.+]] = llvm.extractvalue [[tape]][0, 0] : !llvm.struct<(struct<(struct<(ptr, ptr, i64)>)>)>
+    // CHECK: [[tape0struct:%.+]] = llvm.extractvalue [[tape]][0] : !llvm.struct<(struct<(ptr, ptr, i64)>)>
     // CHECK: [[tape0memref:%.+]] = builtin.unrealized_conversion_cast [[tape0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
     // CHECK: [[result0memref:%.+]] = call @rev([[tape0memref]], [[cotan0memref]])
     %0 = func.call @rev(%arg4, %arg3) : (memref<f64>, memref<f64>) -> memref<f64>
```

The test succeeds, but when running, Enzyme does issue an error:

```python
import pennylane as qml

from catalyst import debug, grad, pure_callback, jacobian

import jax
#x = jax.numpy.array([[3.0, 2.0], [1.0, 1.0]])

@pure_callback
def identity(x) -> float:
    return x


@identity.fwd
def fwd(x):
    return identity(x), (0.5, 0.5) # more than 1


@identity.bwd
def bwd(res, cot):
    return cot * (res[0] + res[1])


@qml.qjit(keep_intermediate=True)
@jacobian
def wrapper(x):
    return identity(x)

x = 1.0
print(wrapper(x))
```

```
 trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
args[0] =   %14 = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[1] =   %"'ipa" = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[2] =   %15 = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[3] =   %"'ipa3" = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[4] =   %_augmented = call { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } @fwd.fwd(ptr %14, ptr %"'ipa", ptr %15, ptr %"'ipa3") FT:{ { ptr, ptr, i64 }, { ptr, ptr, i64 } }
```

But without it, it works alright. With an arbitrary number of elements in the tape.

File: mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp (hunk 5 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp:61]: This is also not possible. Because the function to be derived is a catalyst.callback and the function that lowers the GradTriple will generate `func.constant` and `func.constant` can only use `func.func` symbols.

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp (hunk 1 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:260]: https://github.com/PennyLaneAI/catalyst/pull/822/commits/06ccbeee620b7b91365927737dd73f8d49d09349

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp (hunk 2 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp]: https://github.com/PennyLaneAI/catalyst/pull/822/commits/dac41e4dacb51b1930086db5572a9963c2184f37

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: https://github.com/PennyLaneAI/catalyst/pull/822/commits/8411b6228c7bbbb46336d15f12e3336939ded337

**Comment by erick-xanadu** [comment]
@dime10 please approve :)

File: mlir/include/Gradient/IR/GradientOps.td [hunk 6]
```diff
@@ -191,5 +192,128 @@ def VJPOp : Gradient_Op<"vjp", [
     let hasVerifier = 1;
 }
 
+def ForwardOp : Gradient_Op<"forward",
```
Comment (paul0403) [inline mlir/include/Gradient/IR/GradientOps.td:195]: Do the `FowardOp` and `ReverseOp` not need assembly formats?

Comment (erick-xanadu) [reply mlir/include/Gradient/IR/GradientOps.td:195]: It's the opposite, they have their own assembly format. It is declared below:

```
  let hasCustomAssemblyFormat = 1;
```

and defined like this in a different file.

```cpp
void ForwardOp::print(OpAsmPrinter &p)
{
    function_interface_impl::printFunctionOp(p, *this, /*isVariadic=*/false,
                                             getFunctionTypeAttrName(), getArgAttrsAttrName(),
                                             getResAttrsAttrName());
}
```

I'm just reusing the implementation of the function op interface.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp (hunk 4 above)
Comment (dime10) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: The only reason I'm concerned is that we don't understand the convention correctly then, at least from what we discussed yesterday?

Btw I don't know if this is case, but whenever we said yesterday "so and so is because that's what enzyme expects", it would great to explicitly note down the assumption that we're making there in comments/docstrings. So here as well we could use a comment "Enzyme expects the following format for ... : <...>"

File: mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp (hunk 5 above)
Comment (dime10) [reply mlir/lib/Gradient/Transforms/gradient_to_llvm.cpp:61]: Maybe a downside after all to create all these operations which are effectively functions? 😅 Either way I think this is okay for now, the issue is well documented 👍

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp (hunk 4 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: I think the issue is that it is also poorly documented on Enzyme. I have tested this experimentally and it works. I'll add the assumption and try to contact the Enzyme authors and submit a PR to Enzyme next quarter. We can make it a ticket :)

Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: https://github.com/PennyLaneAI/catalyst/pull/822/commits/fa0f276ed784dd56a8440aea41d52da97d93ccd1

Comment (dime10) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: So just to confirm, are you saying that in that example enzyme tried to call the backwards with with the results of the forward pass, on top of the residuals?

Looking at this line I can't really tell, that part looks correct no?
```
trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
```

Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: With the change above, I set the tape of the reverse pass to the one given by the message.

```
trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
```

But if we look at what enzyme is actually passing

```
args[4] =   %_augmented = call { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } @fwd.fwd(ptr %14, ptr %"'ipa", ptr %15, ptr %"'ipa3")
```

If we look at the Enzyme source:

```
            llvm::errs() << "args[" << i << "] = " << *pre_args[i]
                         << " FT:" << *FT->getParamType(i) << "\n";
```

We see that the expected value of the parameter is

```
FT:{ { ptr, ptr, i64 }, { ptr, ptr, i64 } }
```

But it differs from the argument we are sending. In other words:

sending ` { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } ` is what enzyme expects

Comment (dime10) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp]: Thanks Erick, that's really helpful!

**Review by dime10 — APPROVED** [review]
🎉
